backtrace-library/src/androidTest/java/backtraceio/library/BacktraceClientMetricsTest.java

```diff
@@ -0,0 +1,950 @@
+
+        assertFalse(backtraceClient.metrics.addUniqueEvent(null));
+        // Account for the startup unique event
+        assertEquals(1, backtraceClient.metrics.getUniqueEvents().size());
+    }
+
+    @Test
+    public void doNotAddUniqueEventEmptyString() {
+        backtraceClient.metrics.enable(new BacktraceMetricsSettings(universeName, token, defaultBaseUrl, 0));
+
+        assertFalse(backtraceClient.metrics.addUniqueEvent(""));
+        // Account for the startup unique event
+        assertEquals(1, backtraceClient.metrics.getUniqueEvents().size());
+    }
+
+    @Test
+    public void uniqueAttributesPerEventDoNotMutate() {
+        backtraceClient.metrics.enable(new BacktraceMetricsSettings(universeName, token, defaultBaseUrl, 0));
+
+        String expectedKey = "foo";
+        String expectedValue1 = "bar";
+        String expectedValue2 = "baz";
+        backtraceClient.attributes.put(expectedKey, expectedValue1);
+        assertTrue(backtraceClient.metrics.addUniqueEvent(uniqueAttributeName[0]));
+
+        backtraceClient.attributes.put(expectedKey, expectedValue2);
+        assertTrue(backtraceClient.metrics.addUniqueEvent(uniqueAttributeName[1]));
+
+        assertEquals(3, backtraceClient.metrics.getUniqueEvents().size());
+        Event event2 = backtraceClient.metrics.getUniqueEvents().getLast();
+        backtraceClient.metrics.getUniqueEvents().removeLast();
+        Event event1 = backtraceClient.metrics.getUniqueEvents().getLast();
+
+        assertEquals(expectedValue1, event1.getAttributes().get(expectedKey));
+        assertEquals(expectedValue2, event2.getAttributes().get(expectedKey));
+    }
+
+    @Test
+    public void uniqueEventWithCustomAttributeExistsEvenIfCustomAttributeDeletedLater() {
+        backtraceClient.metrics.enable(new BacktraceMetricsSettings(universeName, token, defaultBaseUrl, 0));
+
+        String expectedKey = "foo";
+        String expectedValue = "bar";
+        backtraceClient.attributes.put(expectedKey, expectedValue);
+        assertTrue(backtraceClient.metrics.addUniqueEvent(expectedKey));
+
+        backtraceClient.attributes.remove(expectedKey);
+
+        assertEquals(expectedValue, backtraceClient.metrics.getUniqueEvents().getLast().getAttributes().get(expectedKey));
+        assertEquals(expectedKey, backtraceClient.metrics.getUniqueEvents().getLast().getName());
+    }
+
+    @Test
+    public void uniqueEventUpdateTimestamp() {
+        backtraceClient.metrics.enable(new BacktraceMetricsSettings(universeName, token, defaultBaseUrl, 0));
+
+        assertTrue(backtraceClient.metrics.addUniqueEvent(uniqueAttributeName[0]));
+
+        assertEquals(uniqueAttributeName[0], backtraceClient.metrics.getUniqueEvents().getLast().getName());
+        long previousTimestamp = backtraceClient.metrics.getUniqueEvents().getLast().getTimestamp();
+
+        try {
+            sleep(1000);
```

Comment:
why we have this sleep(..)?

Reply:
Here we have to wait 1 second so that the timestamp will update on the next `send()`. Timestamp granularity is 1 second, I will add a comment to explain this.